doc/MiniLanguageTutorial.md

```diff
@@ -480,37 +562,278 @@ Mini never automatically converts types to make an operation succeed.  Programme
+> A reference to a specialization of a generic function. *ident* must be a generic function name, 
+> *type1*, *type2*, ... must all be types,
+> and there must be the same number of *type*s specified as there are type variables on *ident*.
+> The function reference is treated as the type resulting from substitution of each type variable with the concrete types listed in *type1*, *type2* etc.
+ 
+`error`
+
+> Causes the system to error, moving to the error handler. This expresion returns type *every*.
+
+`"` [a-zA-Z0-9_ .,:?'<+>()!@#$%^&*|~\\/-]* `"`
+
+> Creates a string literal. The type of this string literal is `string`.
+
+`"` (`0x`)? [a-fA-F0-9]* `"`
+
+> Hex string literal, of type `string`.
+
+`{` (*statement*)* (*expression*)? `}`
+
+> Codeblock expression. The *statement*s are executed in order, 
+> all locals defined within the codeblock are only valid within the codeblock.
+> If *expression* is present, the result of *expression* with equivalent type. 
+> Otherwise, the codeblock returns *void*.
+
+`(` *expression* `)`
 
-> Escape to assembly code.  The arguments (*expression1*, *expression2*, etc.), if any, are pushed onto the AVM stack (with *expression1* at the top of the stack). Then the *instructions*, which are a sequence of AVM assembly instructions, are executed.  The assembly instructions are assumed to consume the arguments and leave on the stack a single value of type *type*, which becomes the result of this expression. (There is another form of `asm`, which produces no result value and is a statement.)
+> Equivalent to *expression*.
 
+## Path syntax
 
+*ident* (`::` *ident*)*
 
+> This represents a path with the leftmost identifier being the widest scope.
+> In particular, a valid path may contain a file name followed by a identifier,
+> or it may be a library specifier, either `std` or `core`, followed by a file name and identifier.
+> In the case of a file name followed by a identifier, the file name represents the mini file within the source folder with that name.
+> So if `"arb_os"` is the compile target `name` would represent `arb_os/name.mini`, and the identifier would represent the function or type of the same name in that file.
+> If it starts with a library prefix, the base folder is determined by the prefix, with
+> `std` corresponding to the `std` folder and `core` corresponding to the `builtin` folder.
+> The second and third parts of the path work the same as mentioned previously.
+> A path represented in this format will be called a virtual path, to distinguish it from paths in use by the filesystem.
+
+## Shadowing
+
+In mini, new variables can be created with the same name as old variables, this is called shadowing.
+To prevent ambiguity when dealing with shadowed variables, a reference to the shadowed name is treated as referring to
+the most recently declared variable that is currently in scope. So if a variable is declared inside a codeblock with the same name as another variable outside the codeblock,
```

Comment:
As this is not valid mini syntax its difficult for me to exactly guess your intention, but if `:=` corresponds to `let`, then yes it would shadow `foo` and not modify the outer `foo`, if it is not a `let`, then it would not be a declaration of a variable so it wouldn't apply.

As an example, this program returns `3`:
```
func main() -> any {
    let foo = 3;
    {
        f(3);
        let foo = 5;
    };
    return foo;
}

func f(y: uint) {
    return;
}
```